opentelemetry-exporter-cloud-trace/tests/test_cloud_trace_exporter.py

```diff
@@ -182,28 +185,102 @@ def test_extract_status(self):
             Status(details=None, code=2, message="error_desc"),
         )
 
-    def test_extract_attributes(self):
+    def test_extract_empty_attributes(self):
         self.assertEqual(
             _extract_attributes({}, 4), ProtoSpan.Attributes(attribute_map={})
         )
+
+    def test_extract_variety_of_attributes(self):
         self.assertEqual(
             _extract_attributes(self.attributes_variety_pack, 4),
             self.extracted_attributes_variety_pack,
         )
-        # Test ignoring attributes with illegal value type
+
+    def test_ignore_invalid_attributes(self):
         self.assertEqual(
-            _extract_attributes({"illegal_attribute_value": dict()}, 4),
-            ProtoSpan.Attributes(attribute_map={}, dropped_attributes_count=1),
+            _extract_attributes(
+                {"illegal_attribute_value": dict(), "legal_attribute": 3}, 4
+            ),
+            ProtoSpan.Attributes(
+                attribute_map={"legal_attribute": AttributeValue(int_value=3)},
+                dropped_attributes_count=1,
+            ),
         )
 
+    def test_too_many_attributes(self):
         too_many_attrs = {}
         for attr_key in range(5):
             too_many_attrs[str(attr_key)] = 0
         proto_attrs = _extract_attributes(too_many_attrs, 4)
         self.assertEqual(proto_attrs.dropped_attributes_count, 1)
 
-    def test_extract_events(self):
+    def test_add_agent_attribute(self):
+        self.assertEqual(
+            _extract_attributes({}, 4, add_agent_attr=True),
+            ProtoSpan.Attributes(
+                attribute_map={"g.co/agent": self.agent_code},
+                dropped_attributes_count=0,
+            ),
+        )
+        # Drop existing attributes in favor of the agent attribute
+        self.assertEqual(
+            _extract_attributes(
+                {"attribute_key": "attr_value"}, 1, add_agent_attr=True
+            ),
+            ProtoSpan.Attributes(
+                attribute_map={"g.co/agent": self.agent_code},
+                dropped_attributes_count=1,
+            ),
+        )
+
+    def test_extract_empty_events(self):
         self.assertIsNone(_extract_events([]))
+
+    def test_too_many_events(self):
+        time_in_ns = 1589919268850900051
+        time_in_ms_and_ns = {"seconds": 1589919268, "nanos": 850899968}
```

Comment:
Should these numbers match? `time_in_ns` ends with `900051` but `time_in_ms_and_ns["nanos"]` ends with `899968`.

Reply:
Seems we're run into a bit of numerical uncertainty

Reply:
This is addressed in PR #42.